wnfs/src/private/hamt/node.rs

```diff
@@ -376,76 +377,81 @@ where
-        K: DeserializeOwned + Clone + AsRef<[u8]>,
-        V: DeserializeOwned + Clone,
+        K: DeserializeOwned + Clone + AsRef<[u8]> + 'a,
+        V: DeserializeOwned + Clone + 'a,
+        H: 'a,
     {
-        let bit_index = hashnibbles.try_next()?;
+        Box::pin(async move {
+            let bit_index = hashnibbles.try_next()?;
 
-        // If the bit is not set yet, return None.
-        if !self.bitmask[bit_index] {
-            return Ok((Rc::clone(self), None));
-        }
-
-        let value_index = self.get_value_index(bit_index);
-        Ok(match &self.pointers[value_index] {
-            Pointer::Values(values) => {
-                let mut node = (**self).clone();
-                let value = if values.len() == 1 {
-                    // If the key doesn't match, return without removing.
-                    if &H::hash(&values[0].key) != hashnibbles.digest {
-                        return Ok((Rc::clone(self), None));
-                    }
-                    // If there is only one value, we can remove the entire pointer.
-                    node.bitmask.set(bit_index, false);
-                    match node.pointers.remove(value_index) {
-                        Pointer::Values(mut values) => Some(values.pop().unwrap()),
-                        _ => unreachable!(),
-                    }
-                } else {
-                    // Otherwise, remove just the value.
-                    let mut values = (*values).clone();
-                    values
-                        .iter()
-                        .position(|p| &H::hash(&p.key) == hashnibbles.digest)
-                        .map(|i| {
-                            let value = values.remove(i);
-                            node.pointers[value_index] = Pointer::Values(values);
-                            value
-                        })
-                };
-
-                (Rc::new(node), value)
+            // If the bit is not set yet, return None.
+            if !self.bitmask[bit_index] {
+                return Ok((self, None));
             }
-            Pointer::Link(link) => {
-                let child = Rc::clone(link.resolve_value(store).await?);
-                let (child, value) = child.remove_value(hashnibbles, store).await?;
-
-                let mut node = (**self).clone();
-                if value.is_some() {
-                    // If something has been deleted, we attempt toc canonicalize the pointer.
-                    if let Some(pointer) =
-                        Pointer::Link(Link::from(child)).canonicalize(store).await?
-                    {
-                        node.pointers[value_index] = pointer;
+
+            let value_index = self.get_value_index(bit_index);
+
+            let mut node = Rc::try_unwrap(self).unwrap_or_else(|rc| (*rc).clone());
+
+            let removed = match &mut node.pointers[value_index] {
+                Pointer::Values(values) => {
+                    if values.len() == 1 {
```

Comment:
this nest of nested matches and conditionals is a bit gnarly? Anyway, to break these up?

Reply:
Hm. Hard to do. The most straightforward idea would be to extract all this functionality into the `Pointer` type, but then we'd have mutually recursive async functions, which doesn't really work.

Instead I've opted to merge one level of the `if` into the match, I like that version a little better at least.

Reply:
me too.